Controller Plugin framework

markturansky opened the pull request:
This is a sketch for discussion.  I'm curious to know if this is a desirable direction for kube-controller-manager.

Goals:
1.  Change controllers into plugins.  Each is similar in implementation.  Make a framework to run many plugins in kube-controller-manager.  Plugin arch copied from VolumePlugin and NetworkPlugin.
2.  Be ready to support @lavalamp's Controller Framework (#5270)
3.  Port 1 controller (i used EndpointsController) as an example.
4.  Future enhancements:
   1. Potentially change kube-controller-manager into a server to allow queries into its status.  The manager can query each of its controllers and report back on each of them.
   2. Allow graceful shutdown of all controllers via stop channels (Smith's Controller Framework has a similar stop mechanism)
   3. Allow set/override default sync values (defaults set by plugins, override by CLI flags)
   4. Refactor CLI sync flags to allow plugin to add these somehow, as opposed to explicitly adding each.

Optional:  Potentially rename all controllers to something akin to "_Manager" or "_Synchronizer" to differentiate a process from a kind.  Example:  PersistentVolumeManager is the sync loop.  PersistentVolumeController is a proposed Kind in the API that makes new PVs.

Genesis: I wrote a controller in which I thought I needed to reconcile 2 objects and didn't like the duplication required.  It wasn't hard to genericize, after which I thought I'd take a shot at making all controllers work similarly.  Porting endpoints_controller was very easy to do.  

@thockin @bgrant0607

bgrant0607 commented:
We definitely intend to convert replication controller manager to a plugin. https://github.com/GoogleCloudPlatform/kubernetes/issues/3058#issuecomment-69067542

satnam6502 commented on pkg/controller/endpoints/endpoints_mgr.go:
```diff
@@ -0,0 +1,235 @@
+	return endpointsManagerName
+}
+
+func (plugin *endpointsManager) GetDefaultSyncPeriod() time.Duration {
```
Minor, but I think this should be `DefaultSyncPeriod` and likewise for the others below that start with `Get`.

satnam6502 commented on pkg/controller/endpoints/endpoints_mgr.go:
```diff
@@ -0,0 +1,235 @@
+	if err != nil {
+		glog.Errorf("Error syncing service: %s/%s, skipping", service.Namespace, service.Name)
+		resultErr = err
+		return resultErr
```
Why not just `return err` ?

satnam6502 commented on pkg/controller/endpoints/endpoints_mgr.go:
```diff
@@ -0,0 +1,235 @@
+			}
+			if len(pod.Status.PodIP) == 0 {
+				glog.Errorf("Failed to find an IP for pod %s/%s", pod.Namespace, pod.Name)
+				return resultErr
```
Where is `resultErr` set?

satnam6502 commented on pkg/controller/endpoints/endpoints_mgr.go:
```diff
@@ -0,0 +1,235 @@
+	service := obj.(*api.Service)
+	glog.Infof("Synchronizing: %s", service.Name)
+
+	var resultErr error
```
At first glance I can't understand the use of `resultErr` in this function.

satnam6502 commented on pkg/controller/plugins.go:
```diff
@@ -0,0 +1,130 @@
+
+type ControllerHost interface {
+	// GetKubeClient returns a client interface
+	GetKubeClient() client.Interface
```
No `Get`.

markturansky replied:
@satnam6502 It looks like VolumeHost and Host (for NetworkPlugins) use Get while the plugins themselves do not (VolPlug, NetworkPlug).  Should I be consistent here or should the others change to follow the Go conventions you linked to in another comment?

satnam6502 replied:
I vote for trying our best to follow the Go style guidelines i.e. don't use `Get` here and look to change our other uses of getters to remove the `Get` prefix.

markturansky replied:
@thockin what do you think?  you made the volume plugins.

satnam6502 commented on pkg/controller/plugins.go:
```diff
@@ -0,0 +1,130 @@
+		return plugin, nil
+	}
+
+	return nil, fmt.Errorf("no controller plugin matched")
```
Perhaps add `name` to the error message?

satnam6502 commented:
I am on vacation for most of this week and for the rest of the time I have to focus on some non-Kubernetes work so I would prefer it if someone else reviewed this to avoid holding up this PR since I don't have the time in the coming days to do it justice.

markturansky commented:
@thockin @lavalamp PTAL?  

1.  I refactored my plugins to use @lavalamp's new controller framework, including the graceful termination of all plugins via a stop channel.
2.  A Controller can sync many objects, not just one.  The Controller itself runs many of @lavalamp's control loops.
3.  The unit test and working example watches 2 objects.
4.  I removed Endpoints entirely to avoid distraction.  The unit test has a fully working example.
5.  I'd like to re-implement the PersistentVolumeClaimBinder in this framework.

lavalamp commented:
I will try to take a look at this tomorrow.

thockin commented:
Is this still RFC or should we update the title?

markturansky commented:
Since I haven't gotten shotdown for turning controllers into plugins, I'll rename to WIP :)

thockin commented on pkg/controller/controller.go:
```diff
@@ -0,0 +1,49 @@
+	"time"
+)
+
+// ControllerPlugin is an interface to controller plugins that can be run
```
I would put this definition in plugins.go.

Optional: Consider just calling it Plugin, which is more go-idiomatic.  Volumes are named as such to disambiguate between multiple kinds of plugin.  I don't think we need that here.

markturansky replied:
Moved and renamed.

NetworkPlugin doesn't follow this pattern, though its Host is named simply "Host".

thockin commented on pkg/controller/controller.go:
```diff
@@ -0,0 +1,49 @@
+	Name() string
+
+	// DefaultSyncPeriod gets the suggested sync period for the plugin.
+	// The default period is not guaranteed to be honored.
```
Can you say more about why it might not be honored, and what bounds the plugin-author can assume?  Example, is this an upper bound on frequency, or a lower bound or neither?  What *are* the bounds?  What if I don't want my plugin running too fast or too slow?

thockin commented on pkg/controller/controller.go:
```diff
@@ -0,0 +1,49 @@
+// Controller represents a long-lived process that watches and synchronizes objects in the system.
+// Controllers watch any number of objects for changes, with one Config for each watch/process
+type Controller interface {
+	ControllerPlugin
```
It's a little weird to embed the plugin into the instance.  For example, this is now legal:

``
plugin.CreateController().CreateController().CreateController().Init(host)
```

If you want some of the same methods, define them explicitly.

markturansky replied:
Right, not sure why I embedded that there.  It is unneeded.

thockin commented on pkg/controller/controller.go:
```diff
@@ -0,0 +1,49 @@
+// Controllers watch any number of objects for changes, with one Config for each watch/process
+type Controller interface {
+	ControllerPlugin
+	Configs() []*framework.Config
```
Comment.  Is this returning extant state or is it supposed to do heavy lifting?  Is it idempotent?  Should it cache results internally?

thockin commented on pkg/controller/controller_test.go:
```diff
@@ -0,0 +1,238 @@
+
+func (plugin *testControllerPlugin) Configs() []*framework.Config {
+	// source simulates an apiserver object endpoint.
+	plugin.pvSource = framework.NewFakeControllerSource()
```
This triggered my comment above on the semantics of this method.  I expected a function called Configs() to just return me the cached configs and be idempotent, which this is *not*.

thockin commented on pkg/controller/controller_test.go:
```diff
@@ -0,0 +1,238 @@
+	pv := delta.Object.(*api.PersistentVolume)
+	if (delta.Type == cache.Added || delta.Type == cache.Sync) && !plugin.exists(pv.Name) {
+		plugin.actualVolumeAddedCount++
+		fmt.Println(delta.Type, pv.Name, plugin.actualVolumeAddedCount)
```
debug code, I assume?  I like to flag debug code with a trailing //FIXME so I know it's debug

thockin commented on pkg/controller/controller_test.go:
```diff
@@ -0,0 +1,238 @@
+
+	delta := obj.(cache.Deltas).Oldest()
+	pv := delta.Object.(*api.PersistentVolume)
+	if (delta.Type == cache.Added || delta.Type == cache.Sync) && !plugin.exists(pv.Name) {
```
What if type == Sync and plugin.exists() ?  You just lose the update?

markturansky replied:
I refactored the code a bit and added a comment that explains updates/syncs were ignored for the purpose of this test.

thockin commented on pkg/controller/controller_test.go:
```diff
@@ -0,0 +1,238 @@
+		ObjectType:       &api.PersistentVolume{},
+		FullResyncPeriod: 5 * time.Second,
+		RetryOnError:     true,
+		Process:          plugin.processPersistentVolume,
```
I see us doing a lot of function pointers lately, and it seems mostly to be a way to dodge adding a typedef.  I am not the deepest in Go, but it seems that it would be more Goan (Goish?  Goy?) to do

```
package framework
type Config interface {
    GetQueue() whatever
    GetListerWatcher() whatever
    ...
    Process()
}

type StaticConfig struct {
    Queue 
    Listerwatcher
   ...
}

func (sc *StaticConfig) GetQueue() { return sc.Queue }
```

and

```
type myConfig struct {
    framework.StaticConfig
}

func (mc *myConfig) Process() { ...}
```


Upon typing that out, I realize how tedious it is, but I want to raise the point anyway.  There's an inflection point somewhere, whereupon adding function pointers is less readable that this mess..

thockin commented on pkg/controller/plugins.go:
```diff
@@ -0,0 +1,144 @@
+)
+
+// ControllerHost is an interface that plugins can use to access the kube-controller-manager
+type ControllerHost interface {
```
As with Plugin, consider naming this just Host for the same reasons cited above

markturansky replied:
Renamed.

thockin commented on pkg/controller/plugins.go:
```diff
@@ -0,0 +1,144 @@
+	// An error is returned if a period does not exist for the given name.
+	SyncPeriod(name string) (*time.Duration, error)
+
+	// SetSyncPeriod allows the plugin to set a default period with the host, also allows the host to override.
```
How does the host override?  By just not accepting this new value and returning its own value in SyncPeriod?  Maybe this should return bool or error if the host does not accept the value?  Or can the host vary the value dynamically (and therefore have no hook to tell the plugin) ?  Once you decide, comments!

thockin commented on pkg/controller/plugins.go:
```diff
@@ -0,0 +1,144 @@
+}
+
+// ControllerPluginManager tracks and manages the lifecycle of registered plugins
+type ControllerPluginManager struct {
```
As above, PluginManager.  In Volumes I went with Mgr, which I think is pretty universally understood.  we should be consistent

thockin commented on pkg/controller/plugins.go:
```diff
@@ -0,0 +1,144 @@
+	return nil, fmt.Errorf("no controller plugin found matching %s", name)
+}
+
+// RunAll runs all Controllers in its own Go routine
```
nit: "runs each Controller in its own goroutine" or "runs all Controllers in their own goroutines"

goroutine, not "Go routine"

thockin commented:
I like it, but there are a lot of loose details that should be pinned down before we tell other people to code to this API.  The bar for documentation and corner-cases of "should be opaque" plugin interfaces is higher than plain-old-code.

derekwaynecarr commented:
Is there an assumption that the only command line argument supported by a controller plugin would be a sync period?  I am not sure that will always be the case, so I am a little wary on hard-coding that as the lone type of argument supported.

From a plugin model, I would like an easy way to say this controller daemon should only run this plugin or set of plugins.  Is that captured somewhere I missed?  (Need to go back and look...)

Sent from my iPhone

> On Apr 13, 2015, at 5:46 PM, Tim Hockin <notifications@github.com> wrote:
> 
> I like it, but there are a lot of loose details that should be pinned down before we tell other people to code to this API. The bar for documentation and corner-cases of "should be opaque" plugin interfaces is higher than plain-old-code.
> 
> —
> Reply to this email directly or view it on GitHub.
>

derekwaynecarr commented:
Other comment I have is around controller host, right now all controllers share the same Kube client but longer term I think they need to use different clients with different authorities scoped to the controller.  Also sharing a single Kube client makes how we could better scope the rate limiting around various Kube clients more or difficult.  Have you thought some about if we went this model we could get clients with different authority?  I know we have this problem in OpenShift and @liggitt has been asking for us to be very explicit about what each client in each controller actually needs to do so he can lock each of the down.  Imagine we will need the same in KUBERNETES longer term too.

markturansky commented:
@derekwaynecarr 

Last comment first... It's easy to make a PluginMgr and call InitPlugins([] plugins), so it is easy to imagine a larger framework running many of these under a security model.  I don't seek to solve that with this first pass at plugins but the path towards it is clear.

Regarding the controller daemon, the kube-controller-manager binary would compile in its set of controller plugins, like Kubelet compiles its volume plugins.  I removed that bit of code when I made the unit test.  That would be the last step, I think, before this goes live.

As for the CLI, if you're advocating for multiple-argument controllers, I agree.  All the other controllers are simple loops.  NodeController is a different animal.  That one would require more configuration parameters than the rest.  I haven't put much thought into this yet, but I agree  a plugin should add its configurable properties to the CLI.

thockin commented:
@derekwaynecarr 

> Is there an assumption that the only command line argument supported by a controller plugin would be a sync period?  I am not sure that will always be the case, so I am a little wary on hard-coding that as the lone type of argument supported.

We can always expand params later.  It's not like this is a user-facing API.  We control the code and can fix plugins.

> From a plugin model, I would like an easy way to say this controller daemon should only run this plugin or set of plugins.  Is that captured somewhere I missed?  (Need to go back and look...)

I don't think that's part of the plugin interface itself, but more the plugin host.  And it seems reasonable to make the host have a --plugins_whitelist or --plugins-blacklist or something.

thockin commented:
@derekwaynecarr 

Regarding clients, I guess it's reasonable that we might change from a host.GetKubeClient() to host.GetKubeClient(myplugin_name) or even pass it in at construction.  The creds could be discovered by secrets or by files on disk or something.  Plugins would never know what happened, just that they got a client and it works.

liggitt commented:
> Plugins would never know what happened, just that they got a client and it works.

That's fine, it would just be good for controllers/plugins to clearly document the permissions it needs, so a client with restricted permissions can be given to it without just trial/error

thockin commented:
Status on this effort?

markturansky commented:
@thockin I will be addressing all this feedback and continue this effort, but my focus has been getting my own controller ready to go (https://github.com/GoogleCloudPlatform/kubernetes/pull/6105)

I'd like to port my controller to this framework, but the former is the needful at this point.

bgrant0607 commented:
cc @davidopp @rjnagal @bprashanth 

May be useful for Deployment and Job.

markturansky commented:
Current state:  wired into the cmd line flags and things look good overall but there are two bugs, both related to scoping and/or pointers.

The plugins can add their command line flags, but the pointer to the value in a map storing them is incorrect and PFlag is not setting them.

The test has a scoping issue with the event handler funcs.  They are not being called to handle an event.

lavalamp commented on pkg/controller/controller.go:
```diff
@@ -0,0 +1,30 @@
+// Controller represents a long-lived process that watches objects in the system and can respond to events.  Plugins
+// implement factory methods for creating framework.Controller instances, such as those created by framework.NewInformer()
+// or framework.New(config).
+type Controller interface {
```
ControllerLister

markturansky replied:
@lavalamp can you explain this a little for me?  Why is it a "lister"?

I think the top level thing is a "manager" or sorts, like ReplicationControllerManager or PersistentVolumeManager.  Each thing can have many Controllers (your framework) within it, but it is still 1 thing managed by the PluginMgr.

Overloading the term "Controller" is problematic, I think.  Any suggestions for clearing up the ambiguity?

lavalamp replied:
I misunderstood what you were trying to do a little bit-- I thought you were using this type to represent a list of plugins, not a list of controllers that compose a single plugin.

So now that I understand what you were trying to do, my comment is: don't help people write their controller, because a list of framework.Controllers isn't a good model for that. Instead, define an interface for the Run() method or whatever, so that people can implement it and have their controller be a plugin in your plugin system.

markturansky replied:
Understood and will change.  Idempotent Run, Status, and graceful Stop seem appropriate.

lavalamp replied:
SGTM, thanks!

lavalamp replied:
(consider also using the ```stopch <-chan struct{}``` pattern for stopping)

markturansky replied:
I think that would be left to the controller itself now, correct?  I can write it in the docs/comments and we enforce that through review.

If the pluginMgr were doing the running, I would change it to what you suggest, but that piece will no longer keep the stop channel.

lavalamp commented on pkg/controller/controller.go:
```diff
@@ -0,0 +1,30 @@
+type Controller interface {
+	// Controllers returns many framework.Controller objects completely configured and ready to run
+	// This function is called once by the framework but should still be idempotent.
+	Controllers() ([]*framework.Controller, error)
```
Please define a new interface type in this package that includes only the methods you care about (e.g. Run()).

lavalamp replied:
This is important because I don't want to conflate the implementation with the interface; many entire controllers involve more than one framework.Controller.

markturansky replied:
@lavalamp I admit this naming is clunky, but I think correct (though we can find a better name).

The PluginMgr will run all the controllers, keep their stop channels, be able to report status, allow graceful shutdown, etc.  I don't think the controllers themselves need Run.  The plugins need to return N controllers, as you say, because an entire controller may contain many control loops.  

This ```Controllers()``` func is more like volume.NewBuilder or NewCleaner.   It is a factory meant to return instances from the plugin, which the plugin would then run.  PluginMgr reduces the duplication of every controller having their own Run method.

lavalamp replied:
I don't think that's sufficient to cover our current controllers. Take a look at pkg/service/endpoints_controller.go's Run() method. It runs two controllers, kicks off a cleanup method, and starts some worker queue draining thing. Please don't solve the problem of freeing controllers from writing their own Run() method; that's a feature, not a bug. :)

markturansky replied:
Thanks for clarifying  :)

lavalamp commented:
As a pre-req to this change, please move the replication controller specific code out of pkg/controller (to e.g. pkg/controller/replication or some such).

davidopp commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -100,7 +102,13 @@ func NewCMServer() *CMServer {
 	}
+	s.PluginMgr.InitFlags(ProbeControllerPlugins())
+	t := 5 * time.Second
+	s.PluginMgr.SyncPeriods["pvclaimbinder-sync-period"] = &t
```
is there any reason you need to use all lower case here? pvclaimbinder is very mysterious unless you already know which four letters are supposed to be capitalized.

davidopp replied:
or better yet, just spell it out (persistent-volume-claim-binder-sync-period)

lavalamp replied:
Also why does this store a pointer to a duration? Shouldn't just a Duration be fine?

lavalamp replied:
nvm I see this is for flags...

bgrant0607 commented:
@markturansky Are any critical changes dependent on this, or can it wait?

markturansky commented:
@bgrant0607 This can wait.

markturansky committed:
Added Controller plugins and ControllerConfig

markturansky commented:
@thockin @lavalamp Controller plugins ready for review.   The new VolumeConfig was mimicked in this PR (and I removed all the other cruft I added).  

Controller plugins now just have Start, Stop, and Status.  I also added Fakes for testing, just like volumes.

I'd like to make the reference implementation the new controller I'm going to add for provisioning.

markturansky commented:
@k8s-bot test this

markturansky commented:
Just 6 files and all green on the builds ...

bgrant0607 commented:
cc @davidopp

bgrant0607 commented:
@k8s-bot test this

markturansky commented:
@k8s-bot test this

lavalamp commented on pkg/controller/testing.go:
```diff
@@ -0,0 +1,77 @@
+limitations under the License.
+*/
+
+package controller
```
Move to a separate fake package?

markturansky replied:
I can, but this follows the pattern in pkg/volume.

lavalamp replied:
This is part of raising the code quality bar. Test-only code should not be in non-test-only packages. We're guilty of it all over the place and we need to stop.

markturansky replied:
Any other test package in the code can easily include the controller/testing pkg for the Fakes.  I'll move this.

lavalamp commented on pkg/controller/plugins.go:
```diff
@@ -0,0 +1,228 @@
+// in kube-controller-manager.  A PluginMgr handles instantiating
+// and managing the lifecycle of controllers.
+type Plugin interface {
+	// Init initializes the plugin.  This will be called once before Run.
```
what is Host for?

markturansky replied:
Host is the connection between the running environment and the plugin.   This follows the plugin pattern for volume and network plugins.

lavalamp commented on pkg/controller/plugins.go:
```diff
@@ -0,0 +1,228 @@
+	Name() string
+
+	// NewController returns an instance of this plugin's controller.
+	NewController() (Controller, error)
```
I do not expect to see Init and NewController in the same thing. NewController makes me think this is a factory, but why would a factory have an Init method?

markturansky replied:
The plugin has to create instances of a Controller (not your framework controllers but whatever this plugin is).

lavalamp replied:
You'll only run one at a time, though, no? Why not just have Start() and
Stop() methods in that case?

On Mon, Sep 14, 2015 at 1:51 PM, Mark Turansky <notifications@github.com>
wrote:

> In pkg/controller/plugins.go
> <https://github.com/kubernetes/kubernetes/pull/6520#discussion_r39445299>:
>
> > +)
> > +
> > +// Plugin is an interface to controller plugins that can be run
> > +// in kube-controller-manager.  A PluginMgr handles instantiating
> > +// and managing the lifecycle of controllers.
> > +type Plugin interface {
> > +	// Init initializes the plugin.  This will be called once before Run.
> > +	Init(host Host)
> > +
> > +	// Name returns the plugin's name.  Plugins should use namespaced names
> > +	// such as "example.com/replication-controller".  The "kubernetes.io" namespace is
> > +	// reserved for plugins which are bundled with kubernetes.
> > +	Name() string
> > +
> > +	// NewController returns an instance of this plugin's controller.
> > +	NewController() (Controller, error)
>
> The plugin has to create instances of a Controller (not your framework
> controllers but whatever this plugin is).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/6520/files#r39445299>.
>

markturansky replied:
I think that's reasonable.  I don't think I need Status in a plugin considering RunAll will receive the runtime error and store status for that individual plugin.

lavalamp replied:
And I don't understand why you would need both Init and Start.

lavalamp replied:
I think you do need a status, because what if the controller suffers a fatal error and stops?

markturansky replied:
ok, so keep the Status func on a Controller.

As for init/start, I can see why Volumes would need init (initialized all plugins once, we'll get many different instances of them later).

Controllers, though, are just instantiated once and run.  

Do we need separate Plugin and Controller interfaces, then?  

instead of plugin.NewController().Start() is it just plugin.Start()?

lavalamp replied:
Yes, that's what I'm suggesting. Too many layers here. :)

On Mon, Sep 14, 2015 at 7:40 PM, Mark Turansky <notifications@github.com>
wrote:

> In pkg/controller/plugins.go
> <https://github.com/kubernetes/kubernetes/pull/6520#discussion_r39469433>:
>
> > +)
> > +
> > +// Plugin is an interface to controller plugins that can be run
> > +// in kube-controller-manager.  A PluginMgr handles instantiating
> > +// and managing the lifecycle of controllers.
> > +type Plugin interface {
> > +	// Init initializes the plugin.  This will be called once before Run.
> > +	Init(host Host)
> > +
> > +	// Name returns the plugin's name.  Plugins should use namespaced names
> > +	// such as "example.com/replication-controller".  The "kubernetes.io" namespace is
> > +	// reserved for plugins which are bundled with kubernetes.
> > +	Name() string
> > +
> > +	// NewController returns an instance of this plugin's controller.
> > +	NewController() (Controller, error)
>
> ok, so keep the Status func on a Controller.
>
> As for init/start, I can see why Volumes would need init (initialized all
> plugins once, we'll get many different instances of them later).
>
> Controllers, though, are just instantiated once and run.
>
> Do we need separate Plugin and Controller interfaces, then?
>
> instead of plugin.NewController().Start() is it just plugin.Start()?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/6520/files#r39469433>.
>

markturansky replied:
I understand and will make these changes.

lavalamp commented on pkg/controller/plugins.go:
```diff
@@ -0,0 +1,228 @@
+	NewController() (Controller, error)
+}
+
+// Host is an interface that plugins can use to access the kube-controller-manager
```
Why is it called "host"? Also, it seems to let you talk to apiserver, not controller manager?

markturansky replied:
This also follows the pattern in pkg/volume and also network plugins.

lavalamp replied:
Please expand the comment. I didn't have much to do with those plugins, so referring to them doesn't help me understand this.

lavalamp replied:
And I still feel like Host is a confusing name for this. "Environment"?

markturansky replied:
Can you explain why you think its confusing in this context but not in network or volume plugins?

I tend to feel that keeping the same pattern (which, to me, includes naming) makes the code in each of the areas more understandable.  There's no parsing required to understand what "Foo" means in this context but is "Bar" in another context.

lavalamp commented on pkg/controller/plugins.go:
```diff
@@ -0,0 +1,228 @@
+
+// PluginMgr is an interface to a component that manages the lifecycle of controller plugins.
+type PluginMgr interface {
+	InitPlugins(plugins []Plugin, host Host) error
```
I would prefer to see an interface where plugins can be added one or several at a time instead of requiring them all at once.

markturansky replied:
I can do that.  I can understand how this would differ from pkg/volumes because we intend to have an API for adding controllers, right?   I imagine they'd plug in one at a time.

lavalamp replied:
Additionally, it seems weird to pass host in at the same time as the plugins. Host seems like it should be an argument to the constructor that creates a PluginMgr.

markturansky replied:
I already lost my thread in this long PR, but elsewhere I asked if we just need plugin.Start() instead of plugin.NewController().Start().

If so, would it be plugin.Start(host) ?

The host needs to get set on the plugin somewhere.

lavalamp replied:
Yes, plugin.Start(host) seems good.

On Mon, Sep 14, 2015 at 7:57 PM, Mark Turansky <notifications@github.com>
wrote:

> In pkg/controller/plugins.go
> <https://github.com/kubernetes/kubernetes/pull/6520#discussion_r39470176>:
>
> > +	// reserved for plugins which are bundled with kubernetes.
> > +	Name() string
> > +
> > +	// NewController returns an instance of this plugin's controller.
> > +	NewController() (Controller, error)
> > +}
> > +
> > +// Host is an interface that plugins can use to access the kube-controller-manager
> > +type Host interface {
> > +	// GetKubeClient returns a client interface
> > +	GetKubeClient() client.Interface
> > +}
> > +
> > +// PluginMgr is an interface to a component that manages the lifecycle of controller plugins.
> > +type PluginMgr interface {
> > +	InitPlugins(plugins []Plugin, host Host) error
>
> I already lost my thread in this long PR, but elsewhere I asked if we just
> need plugin.Start() instead of plugin.NewController().Start().
>
> If so, would it be plugin.Start(host) ?
>
> The host needs to get set on the plugin somewhere.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/6520/files#r39470176>.
>

lavalamp commented on pkg/controller/plugins.go:
```diff
@@ -0,0 +1,228 @@
+}
+
+// Status returns the runtime status of all plugins under management
+func (mgr *pluginMgr) Status() []PluginStatus {
```
needs a lock?

lavalamp commented on pkg/controller/plugins.go:
```diff
@@ -0,0 +1,228 @@
+		if managedPlugin.err != nil || !managedPlugin.running {
+			continue
+		}
+		managedPlugin.err = managedPlugin.controller.Stop()
```
move this to be a method on managedPlugin; IMO you shouldn't set an internal field like this.

lavalamp replied:
This goes for everything that sets fields on managedPlugin objects.

lavalamp commented on cmd/kube-controller-manager/app/plugins.go:
```diff
@@ -53,3 +54,16 @@ func ProbeRecyclableVolumePlugins(flags VolumeConfigFlags) []volume.VolumePlugin
+		SyncPeriod: flags.ResourceQuotaSyncPeriod,
+	}
+
+	// allPlugins = append(allPlugins, resourcequota.ProbeControllerPlugins(config)...)
```
I don't understand this

markturansky replied:
This mimics how we added VolumeConfig.  This is here as an example of its usage and would eventually be replaced by real usage as controllers are converted into plugins.

Specific attributes are in the binary as flags, then they are passed generically via the config struct.

lavalamp commented on pkg/controller/plugins.go:
```diff
@@ -0,0 +1,228 @@
+	for name, managedPlugin := range mgr.plugins {
+		status := PluginStatus{
+			Name:    name,
+			Running: managedPlugin.running,
```
This seems potentially deceptive-- you should have a way to ask the plugin what its status is, not just assume that it's running because you started it some amount of time ago.

markturansky replied:
Agreed.  will get this from the Status func

lavalamp commented on pkg/controller/plugins.go:
```diff
@@ -0,0 +1,228 @@
+	mgr.mutex.Lock()
+	defer mgr.mutex.Unlock()
+	for name, managedPlugin := range mgr.plugins {
+		// do not attempt to run a plugin that had previous errors or is already running.
```
So if one plugin has an error the only way to try again is to restart the whole binary?

markturansky replied:
Should the manager attempt periodic restart of a failed controller?

lavalamp replied:
Maybe? Probably? In a rate-limited way?

On Mon, Sep 14, 2015 at 7:55 PM, Mark Turansky <notifications@github.com>
wrote:

> In pkg/controller/plugins.go
> <https://github.com/kubernetes/kubernetes/pull/6520#discussion_r39470080>:
>
> > +func (mgr *pluginMgr) FindPluginByName(name string) (Plugin, error) {
> > +	mgr.mutex.Lock()
> > +	defer mgr.mutex.Unlock()
> > +	if managedPlugin, ok := mgr.plugins[name]; ok {
> > +		return managedPlugin.plugin, nil
> > +	}
> > +	return nil, fmt.Errorf("no controller plugin found matching %s", name)
> > +}
> > +
> > +// RunAll attempts to create Controller instances for each plugin and run them.  This method will not fail
> > +// for any particular plugin.  Runtime Status is kept for all plugins.  Callers should use pluginMgr.Status
> > +func (mgr *pluginMgr) RunAll() {
> > +	mgr.mutex.Lock()
> > +	defer mgr.mutex.Unlock()
> > +	for name, managedPlugin := range mgr.plugins {
> > +		// do not attempt to run a plugin that had previous errors or is already running.
>
> Should the manager attempt periodic restart of a failed controller?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/6520/files#r39470080>.
>

lavalamp commented:
I'm not sure if I see the overall benefit to this. In kubelet, you *have* to have a way of plugging things in, and you have to be able to leverage that to calling out to other binaries when needed. But for controllers, there's no need for that later case, ever-- kube-controller-manager does *not* have to talk to frobber-controller-manager; if you want a 3rd party controller, you can just write one and run it as a separate binary, there's no need for a plug-in mechanism.

I'm not saying I'm against this, either, necessarily.

lavalamp commented on pkg/controller/plugins.go:
```diff
@@ -0,0 +1,228 @@
+limitations under the License.
+*/
+
+package controller
```
This should be in a controller/plugin/manager package, IMO.

markturansky replied:
Cloudprovider, Network, and Volume all have their plugins.go and interfaces in the top level of the pkg.   Why would this one be different?

lavalamp replied:
Well... this one will be different because we'll do it right. :)

More seriously, I'm very focused lately on getting the package dependency
tree into a sane state. This means separating interfaces like this so that
plugin code doesn't need to import manager code and vice versa. The
interfaces need to be in common packages, but no actual code should be.

On Mon, Sep 14, 2015 at 7:55 PM, Mark Turansky <notifications@github.com>
wrote:

> In pkg/controller/plugins.go
> <https://github.com/kubernetes/kubernetes/pull/6520#discussion_r39470075>:
>
> > +Copyright 2015 The Kubernetes Authors All rights reserved.
> > +
> > +Licensed under the Apache License, Version 2.0 (the "License");
> > +you may not use this file except in compliance with the License.
> > +You may obtain a copy of the License at
> > +
> > +    http://www.apache.org/licenses/LICENSE-2.0
> > +
> > +Unless required by applicable law or agreed to in writing, software
> > +distributed under the License is distributed on an "AS IS" BASIS,
> > +WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> > +See the License for the specific language governing permissions and
> > +limitations under the License.
> > +*/
> > +
> > +package controller
>
> Cloudprovider, Network, and Volume all have their plugins.go and
> interfaces in the top level of the pkg. Why would this one be different?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/6520/files#r39470075>.
>

markturansky commented:
@lavalamp thanks for all the feedback.  I see a few good reasons for this PR.  None earth shattering, but I think overall in the right direction.

1.  controllermanager.go becomes smaller and simpler.  It's currently all bespoke initialization of controllers line after line.  Plugins would reduce that dramatically. 

2.  This PR provides a common runtime for all controllers and common start/stop mechanics.  Controllers are already similar to each other.  This would make them more so.

3.  There's no way today of knowing what controllers are running and their health.  The binary can be made into a server (more like kubelet) and it's status could be queried at runtime to view all controllers' health.  Can't do that today.  Might be useful to an admin running a cluster.  Perhaps controllers can collect their own stats and provide them in Status.   

That last one is difficult to do without the common runtime.

lavalamp commented:
Yeah, I think all of those are fine reasons. I guess I'll just point out
that the motivation is different than the motivation for the volume/network
plugins.

On Mon, Sep 14, 2015 at 7:52 PM, Mark Turansky <notifications@github.com>
wrote:

> @lavalamp <https://github.com/lavalamp> thanks for all the feedback. I
> see a few good reasons for this PR. None earth shattering, but I think
> overall in the right direction.
>
>    1.
>
>    controllermanager.go becomes smaller and simpler. It's currently all
>    bespoke initialization of controllers line after line. Plugins would reduce
>    that dramatically.
>    2.
>
>    This PR provides a common runtime for all controllers and common
>    start/stop mechanics. Controllers are already similar to each other. This
>    would make them more so.
>    3.
>
>    There's no way today of knowing what controllers are running and their
>    health. The binary can be made into a server (more like kubelet) and it's
>    status could be queried at runtime to view all controllers' health. Can't
>    do that today. Might be useful to an admin running a cluster. Perhaps
>    controllers can collect their own stats and provide them in Status.
>
> That last one is difficult to do without the common runtime.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/6520#issuecomment-140262285>
> .
>

markturansky commented:
@lavalamp @thockin Is there any appetite for this?  The feedback was lukewarm to begin with.

At the very least, this PR seems low priority.  Close and re-open in the future if there's a need/desire? Add all the suggested feedback and push it through?

sttts commented:
For contrib/mesos we have interest in this to get rid of our controller manager fork which includes quite some code duplication now.

markturansky commented:
Closing this effort.
